### PR TITLE
Test Coverage PR-4: Tests added in ./cmd subdir headers.go and license.go files

### PR DESCRIPTION
**Pull request by mohanmanikanta2299**
## Summary
This PR is the 4th part of the test coverage PRs intended to increase the test coverage of the entire repo from 55% to 80+%, fulfilling our coverage requirements. 

## Changes Made
* Implemented comprehensive table-driven tests for the `cmd` package.
* `license.go`, `headers.go` -- list of files in `cmd` package for which the test have been created/updated.
* Added validation for CLI command flags, ensuring default values and flag bindings are properly configured.
* Added strict assertions (`assert.Equal`, `assert.Contains`) and explicit error checking to validate function outputs and edge cases.
* Used `require` for critical initializations to prevent nil-pointer panics during test execution.

## Testing
* Ran `go test -v ./...` (All tests passing)
* Ran `go test -race ./...` (No race conditions detected)
* Verified coverage metrics locally.

### :link: External Links
https://hashicorp.atlassian.net/browse/CCEN-512

### :thinking: Can be merged upon approval?
:white_check_mark:

## PCI review checklist

- [ ] I have documented a clear reason for, and description of, the change I am making.

- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [ ] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

### CODE REVIEW

File: cmd/headers_test.go [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+
+	buf := new(bytes.Buffer)
+	rootCmd.SetOut(buf)
+	rootCmd.SetErr(buf)
+	headersCmd.SetOut(buf)
+	headersCmd.SetErr(buf)
+	rootCmd.SetArgs([]string{"headers", "--spdx", "MPL-2.0", "--copyright-holder", "Test Corp."})
+
+	_ = rootCmd.Execute()
+
```
[the remark this answers is not in the document]
Comment (mohanmanikanta2299) [reply cmd/headers_test.go:387]: Updated code.

File: cmd/headers_test.go [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+	cmd.Dir = dir
+	require.NoError(t, cmd.Run())
+	cmd = exec.Command("git", "commit", "-m", message)
+	cmd.Dir = dir
+	require.NoError(t, cmd.Run())
+}
+
+// withTestConfig saves the current global conf, applies modifications via fn,
+// and restores the original config when the test completes.
+func withTestConfig(t *testing.T, fn func(c *config.Config)) {
```
Comment (sonamtenzin2) [inline cmd/headers_test.go:22]: In cases the tests are running in parallel, can they read the dirty state of the global config ? If so, it can make the tests brittle

File: cmd/license_test.go [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+	// Use current year for the second commit so the test is stable
+	now := time.Now()
+	dateStr := now.Format("2006-01-02T15:04:05-07:00")
+	cmd = exec.Command("git", "commit", "-m", "second commit", "--date", dateStr)
+	cmd.Dir = tmpDir
+	cmd.Env = append(os.Environ(), "GIT_COMMITTER_DATE="+dateStr)
+	require.NoError(t, cmd.Run())
+
+	oldYear := conf.Project.CopyrightYear
+	defer func() { conf.Project.CopyrightYear = oldYear }()
```
Comment (sonamtenzin2) [inline cmd/license_test.go]: It is preferable to use the t.Cleanup method instead of defer in tests

File: cmd/headers_test.go [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+	"testing"
+	"time"
+
+	"github.com/hashicorp/copywrite/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// initGitRepo initializes a git repository in dir with a dummy commit.
+func initGitRepo(t *testing.T, dir string) {
```
Comment (sonamtenzin2) [inline cmd/headers_test.go]: This and the setupGitRepo can be taken out of theses files and kept as a single util function

File: cmd/headers_test.go [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+			updateLicenseFile(headersCmd, licensePath, true, false)
+
+			content, err := os.ReadFile(licensePath)
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectedContent, string(content))
+		})
+	}
+}
+
+func TestHeadersCmd_PlanMode_NoFiles(t *testing.T) {
```
Comment (sonamtenzin2) [inline cmd/headers_test.go]: nit: Probably change the name to no files updated

File: cmd/headers_test.go [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+	}
+
+	// Verify ignored files were NOT modified (no copyright header added)
+	vendorActual, err := os.ReadFile(vendorFile)
+	require.NoError(t, err)
+	assert.Equal(t, vendorContent, string(vendorActual), "vendor file should not have been modified")
+
+	autogenActual, err := os.ReadFile(autogenFile)
+	require.NoError(t, err)
+	assert.Equal(t, autogenContent, string(autogenActual), "autogen file should not have been modified")
```
Comment (sonamtenzin2) [inline cmd/headers_test.go:403]: Will be better to include a check that a non-ignored file was modified.

File: cmd/license_test.go [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+		if err := rootCmd.Execute(); err != nil {
+			t.Logf("execute: %v", err)
+		}
+		return
+	}
+
+	cmd := exec.Command(os.Args[0], "-test.run=TestLicenseCmd_Run_Plan_MissingLicense", "-test.count=1")
+	cmd.Env = append(os.Environ(), "TEST_LICENSE_PLAN_MISSING=1")
+	output, err := cmd.CombinedOutput()
+	assert.Error(t, err)
```
Comment (sonamtenzin2) [inline cmd/license_test.go]: This might be successful in case any other error also occurs. Better idea will be to check we are getting the error that is expected.

File: cmd/headers_test.go (hunk 6 above)
Comment (mohanmanikanta2299) [reply cmd/headers_test.go:403]: Added 2 cases to test 
- if non ignored file was modified.
- existing correct header is untouched.

File: cmd/headers_test.go (hunk 2 above)
Comment (mohanmanikanta2299) [reply cmd/headers_test.go:22]: The test suite is intentionally written to run sequentially. 
The functions - `updateExistingHeaders`, `updateLicenseFile` and `Cobra` command in `headers.go` file depend on the global `conf` pointer and `os.Chdir` for directory context. SInce it is a CLI tool with such constraints, only one command can run at a time in a single process.

You are right to point out that in case of parallel run, there can be dirty reads possible leading to race condition. The way I can think of achieving success even while running parallel tests is by refactoring the existing production code.

For the current implementation, no parallel calls are present and none should be added as well.

File: cmd/license_test.go [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+	require.NoError(t, err)
+
+	output := buf.String()
+	assert.Contains(t, output, "Validates that a LICENSE file is present")
+	assert.Contains(t, output, "--dirPath")
+	assert.Contains(t, output, "--plan")
+}
+
+func Test_determineLicenseCopyrightYears_WithMultipleCommits(t *testing.T) {
+	tmpDir := t.TempDir()
```
Comment (sonamtenzin2) [inline cmd/license_test.go]: Can this init use the function in helper ? Same for a few more scenarios below

File: cmd/license_test.go [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+	t.Cleanup(func() { licenseCmd.SetOut(origLicOut); licenseCmd.SetErr(origLicErr) })
+
+	buf := new(bytes.Buffer)
+	rootCmd.SetOut(buf)
+	rootCmd.SetErr(buf)
+	licenseCmd.SetOut(buf)
+	licenseCmd.SetErr(buf)
+	rootCmd.SetArgs([]string{"license", fmt.Sprintf("--year=%d", currentYear), "--spdx", "MPL-2.0", "--copyright-holder", "Test Corp.", "--dirPath", "."})
+
+	// Exercises the Run path - may succeed or fail depending on copyright format matching
```
Comment (sonamtenzin2) [inline cmd/license_test.go]: Did not quite get this, why may the test succeed or fail based on the copyright formatting ?

File: cmd/headers_test.go [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+	t.Chdir(tmpDir)
+
+	withTestConfig(t, func(c *config.Config) {
+		c.Project.CopyrightHolder = "Test Corp."
+		c.Project.CopyrightYear = 2023
+		c.Project.IgnoreYear1 = false
+	})
+
+	buf := new(bytes.Buffer)
+	headersCmd.SetOut(buf)
```
Comment (sonamtenzin2) [inline cmd/headers_test.go:101]: Please restore the headersCmd to the original state or we could have non deterministic interactions between tests. Review this for other test cases as well.

Comment (mohanmanikanta2299) [reply cmd/headers_test.go:101]: Added `restoreHeadersCmd` to perform cleanup and restore headersCmd to original state to avoid non-deterministic interactions.

File: cmd/license_test.go (hunk 9 above)
Comment (mohanmanikanta2299) [reply cmd/license_test.go]: Stray comment. Cleaned it up.
Updated it with the relevant comment and also added one more assertion.

File: cmd/license_test.go (hunk 8 above)
Comment (mohanmanikanta2299) [reply cmd/license_test.go]: re-used the helper methods to avoid code redundancy.

**Review by sonamtenzin2 — APPROVED** [review]
